### PR TITLE
Linux jemalloc implementation

### PR DESCRIPTION
**Pull request by arosenberger**
Hi Denys,

This includes a working Linux wrapper for jemalloc. The libscalaOffheap.so statically links the latest jemalloc release, so users do not have that dependency. The .so file is placed in src/main/resources, then unpacked to a temp directory and appended to the java.library.path at runtime. This works in unit tests, but I have not yet created a fat jar and tried from a jar invocation.

Thanks,
Adam

### CODE REVIEW

**Comment by arosenberger** [comment]
Failure from having multiple versions of the mutable suite. Will add a cleanup method and re-submit

File: core/src/c/README.txt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+The makefile assumes a few things
+
+1) Download and extract the jemalloc artifact into /tmp
```
Comment (densh) [inline core/src/c/README.txt]: Is it possible to automate this by automatically cloning and building their latest stable release from https://github.com/jemalloc/jemalloc/tree/master ?

File: core/src/main/scala/offheap/jni/JNILoader.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
+package offheap.jni;
```
Comment (densh) [inline core/src/main/scala/offheap/jni/JNILoader.java]: Our binding strategy is just an implementation detail and thus should rather be put under `offheap.internal` as we won't guarantee stability of this part of API.

File: core/src/c/offheap_jni_JemallocWrapper.cpp [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+#endif
+
+JNIEXPORT jlong JNICALL Java_offheap_jni_JemallocWrapper_malloc_10 (JNIEnv *, jclass, jlong size) {
+    const auto result = je_malloc(static_cast<size_t>(size));
+
+    return result == NULL ? -1 : (jlong) result;
+}
+
+JNIEXPORT jlong JNICALL Java_offheap_jni_JemallocWrapper_realloc_10 (JNIEnv *, jclass, jlong address, jlong newSize) {
+    const auto result = je_realloc((void *) address, static_cast<size_t>(newSize));
```
Comment (densh) [inline core/src/c/offheap_jni_JemallocWrapper.cpp]: Is `size_t` guaranteed to be 64-bit on x86-64? What about x86? 

`s.m.Unsafe` always uses longs for addresses and sizes but it's assumed to be in 32-bit range on 32-bit platforms. I think that we'll need some extra machinery to cast this safely.

**Comment by densh** [comment]
The build fails to compile on JDK 6.

**Comment by densh** [comment]
Also please make sure that you've signed the [Scala CLA](http://www.typesafe.com/contribute/cla/scala).

File: core/src/main/scala/offheap/jni/JNILoader.java (hunk 2 above)
Comment (densh) [inline core/src/main/scala/offheap/jni/JNILoader.java]: Update: `scala.offheap.internal`.

File: core/src/c/offheap_jni_JemallocWrapper.cpp (hunk 3 above)
Comment (arosenberger) [inline core/src/c/offheap_jni_JemallocWrapper.cpp]: gcc will have this as the max value of uint32_t and uint64_t on 32 and 64 bit platforms respectively. So in 64 bit we will be dealing with long's on both sides [and in fact unsigned long on the C++ side] and on 32 bit, the widening of the [assumed] int into an unsigned long won't cause any problems either. I think we are good in both cases.

**Comment by densh** [comment]
Rebased version of this PR now resides in [topic/jemalloc](https://github.com/densh/scala-offheap/commits/topic/jemalloc) branch.

**Pull request closed without merging**